# Exploring Ebay Car Sales Data

In this project, we'll work with a dataset of used cars from *eBay Kleinanzeigen*, a classifieds section of the German eBay website.

The data dictionary provided with data is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included user car listings.

Let's start by importing the libraries we need and reading the dataset into pandas.

In [348]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [349]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [350]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

We know that the dataset has 50,000 rows. Thus, the following columns have null values:

- vehicleType            44905 non-null object
- gearbox                47320 non-null object
- model                  47242 non-null object
- fuelType               45518 non-null object
- notRepairedDamage      40171 non-null object

Besides that, we can also see that:

- the column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores
- the **name** column has long names, such as, *Peugeot_807_160_NAVTECH_ON_BOARD* (maybe we can shorten those names in order to make the analysis easier)
- the **seller**, **offerType**, **vehicleType**, **gearbox**, **fuelType**, **notRepairedDamage** are in German (we'll need to translate them into English)
- the **price** and **odometer** have the object dtype, so we'll need to remove the '$', ',' (comma) and 'km' symbols and convert them into float

Let's start by converting the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [351]:
newColumns = [
    'date_crawled', 
    'name',                   
    'seller',                
    'offer_type',          
    'price',                
    'abtest',               
    'vehicle_type',        
    'registration_year',     
    'gearbox',                
    'power_ps',               
    'model',                 
    'odometer',               
    'registration_month',   
    'fuel_type',               
    'brand',                  
    'unrepaired_damaged',      
    'ad_created',            
    'nr_of_pictures',           
    'postal_code',             
    'last_seen'             
]

autos.columns = newColumns
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damaged', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [352]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damaged           ad_created  nr_of_pictures  \
0     peugeot               nein  2016-03-26 00:00:00               0   
1         bmw               nein  2016-04-04 00:00:00               0   
2  volkswagen               nein  2016-03-26 00:00:00               0   
3       smart               nein  2016-03-12 00:00:00               0   
4        ford               nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [353]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-25 19:57:10  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand  \
count       50000        50000.000000     45518       50000   
unique         13                 NaN         7          40   
top     150,000km                 NaN    benzin  volkswagen   
freq        32424                 NaN     30107       10687   
mean          NaN            5.723360       NaN         NaN   
std           NaN            3.711984       NaN         NaN   
min           NaN            0.000000       NaN         NaN   
25%           NaN            3.000000       NaN         NaN   
50%           NaN            6.000000       NaN         NaN   
75%           NaN            9.000000       NaN         NaN   
max           NaN           12.000000       NaN         NaN   

       unrepaired_damaged           ad_created  nr_of_pictures   postal_code  \
count               40171                50000         50000.0  50000.000000   
unique                  2                   76             NaN           NaN   
top                  nein  2016-04-03 00:00:00             NaN           NaN   
freq                35232                 1946             NaN           NaN   
mean                  NaN                  NaN             0.0  50813.627300   
std                   NaN                  NaN             0.0  25779.747957   
min                   NaN                  NaN             0.0   1067.000000   
25%                   NaN                  NaN             0.0  30451.000000   
50%                   NaN                  NaN             0.0  49577.000000   
75%                   NaN                  NaN             0.0  71540.000000   
max                   NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

From the table above, we can see that there are only 2 values for the following columns and they are candidates to be dropped:

- **seller**
- **offer_type**

The **gearbox** and **unrepaired_damaged** columns also has 2 values, but they may be important for the analysis.

The **nr_of_pictures** column seems to have only zeros.

The **price** and **odometer** columns need to be cleaned and converted into float.

In [354]:
sellers = autos['seller']
offer_types = autos['offer_type']
gearboxes = autos['gearbox']
damaged = autos['unrepaired_damaged']

print(sellers.unique())
print(offer_types.unique())
print(gearboxes.unique())
print(damaged.unique())

['privat' 'gewerblich']
['Angebot' 'Gesuch']
['manuell' 'automatik' nan]
['nein' nan 'ja']


All the values above are in German.

- The seller column has the values 'private' and 'dealer'
- The offer_type column has the values 'offer' and 'request'
- The gearbox column has the values 'manually',  'automatic' and null
- The unrepaired_damaged column has the values 'yes', 'no' and null

In [355]:
pictures = autos['nr_of_pictures']
print(pictures.unique())

[0]


We know now that the nr_of_pictures column has all values zero, so it can be dropped.

In [356]:
prices = autos['price']
print(prices.unique())

['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']


In [357]:
prices = prices.str.replace('$','')
prices = prices.str.replace(',','')
prices = prices.astype(float)
autos['price'] = prices
print(prices.unique())

[ 5000.  8500.  8990. ...   385. 22200. 16995.]


In [358]:
print(prices.unique().shape)
prices.describe()

(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

The mininum (0) and maximum (100,000,000) prices above look unrealistical. 

In [359]:
print(autos['odometer'].unique())

['150,000km' '70,000km' '50,000km' '80,000km' '10,000km' '30,000km'
 '125,000km' '90,000km' '20,000km' '60,000km' '5,000km' '100,000km'
 '40,000km']


In [360]:
autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',','')
autos['odometer'] = autos['odometer'].astype(float)
autos.rename(columns={'odometer' : 'odometer_km'}, inplace=True)
print(autos['odometer_km'].unique())

[150000.  70000.  50000.  80000.  10000.  30000. 125000.  90000.  20000.
  60000.   5000. 100000.  40000.]


In [361]:
print(autos['odometer_km'].describe())

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


The odometer values seem realistic.

Let's keep investigating the price column.

In [362]:
prices.value_counts(ascending=True)

789.0            1
11240.0          1
898.0            1
18310.0          1
6202.0           1
69900.0          1
3129.0           1
310.0            1
5248.0           1
2785.0           1
34940.0          1
345000.0         1
2459.0           1
2001.0           1
46500.0          1
21888.0          1
27020.0          1
5998.0           1
1247.0           1
919.0            1
5180.0           1
10000000.0       1
15911.0          1
21599.0          1
8930.0           1
4780.0           1
69999.0          1
139997.0         1
69997.0          1
2225.0           1
              ... 
1999.0         322
1600.0         327
350.0          335
1250.0         335
5500.0         340
1800.0         355
550.0          356
3000.0         365
1300.0         371
1100.0         376
950.0          379
2200.0         382
300.0          384
4500.0         394
700.0          395
850.0          410
650.0          419
900.0          420
750.0          433
999.0          434
2000.0         460
3500.0      

We can see the price \\$ 0.0 is the most common in the column (1421 entries). Also, there is 1 value of \$ 10,000,000.0. Let's start by removing all zeros and any price above \\$ 10 million.

In [363]:
print(prices.shape)
autos = autos[autos['price'].between(0, 10000000, inclusive=False)]
prices = autos['price']
print(prices.shape)
prices.value_counts(ascending=True)

(50000,)
(48571,)


4349.0        1
18310.0       1
6202.0        1
69900.0       1
3129.0        1
310.0         1
5248.0        1
34940.0       1
919.0         1
345000.0      1
2459.0        1
2001.0        1
46500.0       1
21888.0       1
27020.0       1
5998.0        1
1247.0        1
5180.0        1
898.0         1
11240.0       1
2785.0        1
10790.0       1
8930.0        1
4780.0        1
69999.0       1
139997.0      1
69997.0       1
2225.0        1
4877.0        1
69993.0       1
           ... 
400.0       321
1999.0      322
1600.0      327
350.0       335
1250.0      335
5500.0      340
1800.0      355
550.0       356
3000.0      365
1300.0      371
1100.0      376
950.0       379
2200.0      382
300.0       384
4500.0      394
700.0       395
850.0       410
650.0       419
900.0       420
750.0       433
999.0       434
2000.0      460
3500.0      498
800.0       498
600.0       531
1200.0      639
1000.0      639
2500.0      643
1500.0      734
500.0       781
Name: price, Length: 235

We see that we ended up with 48571 prices, which means that 1429 lines were removed from the dataset (1421 zeros and 8 prices that were above 10 million).

In [364]:
prices.describe()

count    4.857100e+04
mean     6.082245e+03
std      2.278698e+04
min      1.000000e+00
25%      1.200000e+03
50%      3.000000e+03
75%      7.490000e+03
max      3.890000e+06
Name: price, dtype: float64

After removing the outliers, we can see that the new values obtained are more realistic:

- mean: \$ 6,082
- min: \$ 1,000
- max: \$ 3,890,000

Let's move on to the date columns.

In [365]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48571 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          48571 non-null object
name                  48571 non-null object
seller                48571 non-null object
offer_type            48571 non-null object
price                 48571 non-null float64
abtest                48571 non-null object
vehicle_type          43985 non-null object
registration_year     48571 non-null int64
gearbox               46226 non-null object
power_ps              48571 non-null int64
model                 46110 non-null object
odometer_km           48571 non-null float64
registration_month    48571 non-null int64
fuel_type             44539 non-null object
brand                 48571 non-null object
unrepaired_damaged    39468 non-null object
ad_created            48571 non-null object
nr_of_pictures        48571 non-null int64
postal_code           48571 non-null int64
last_seen             48571 non-null object
dtypes: float6

The data columns and their types are:
    
- **date_crawled**          - 48571 non-null object
- **registration_year**     - 48571 non-null int64
- **registration_month**    - 48571 non-null int64
- **ad_created**            - 48571 non-null object
- **last_seen**             - 48571 non-null object

Let's first understand how the values in the three string columns (date_crawled, ad_created and last_seen) are formatted.

In [366]:
autos[['date_crawled', 'ad_created', 'last_seen']][:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We can extract the date values getting the first 10 characters of each string.

In [367]:
date_crawled = autos['date_crawled'].str[:10]
ad_created = autos['ad_created'].str[:10]
last_seen = autos['last_seen'].str[:10]

print(date_crawled.head(), end='\n\n')
print(ad_created.head(), end='\n\n')
print(last_seen.head(), end='\n\n')

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: date_crawled, dtype: object

0    2016-03-26
1    2016-04-04
2    2016-03-26
3    2016-03-12
4    2016-04-01
Name: ad_created, dtype: object

0    2016-04-06
1    2016-04-06
2    2016-04-06
3    2016-03-15
4    2016-04-01
Name: last_seen, dtype: object



Let's calculate the distribution of values in the three columns as percentages and display them by date in ascending order.

In [368]:
date_crawled.value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025324
2016-03-06    0.014041
2016-03-07    0.036009
2016-03-08    0.033291
2016-03-09    0.033086
2016-03-10    0.032180
2016-03-11    0.032571
2016-03-12    0.036915
2016-03-13    0.015668
2016-03-14    0.036544
2016-03-15    0.034280
2016-03-16    0.029606
2016-03-17    0.031644
2016-03-18    0.012909
2016-03-19    0.034774
2016-03-20    0.037883
2016-03-21    0.037389
2016-03-22    0.033003
2016-03-23    0.032221
2016-03-24    0.029338
2016-03-25    0.031603
2016-03-26    0.032200
2016-03-27    0.031089
2016-03-28    0.034856
2016-03-29    0.034115
2016-03-30    0.033683
2016-03-31    0.031830
2016-04-01    0.033683
2016-04-02    0.035474
2016-04-03    0.038603
2016-04-04    0.036524
2016-04-05    0.013094
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The date_crawled values are well distributed between 2016-03-05 and 2016-04-07. It doesn't seem to be relevant to our analysis.

In [369]:
ad_created.value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033147
2016-03-10    0.031891
2016-03-11    0.032900
2016-03-12    0.036750
2016-03-13    0.017006
2016-03-14    0.035186
2016-03-15    0.034012
2016-03-16    0.030121
2016-03-17    0.031294
2016-03-18    0.013588
2016-03-19    0.033683
2016-03-20    0.037944
2016-03-21 

We can see that few ads were created in 2015 and at the beginning of 2016. From March of 2016 on, the values are better distributed.

In [370]:
last_seen.value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005394
2016-03-08    0.007412
2016-03-09    0.009594
2016-03-10    0.010665
2016-03-11    0.012374
2016-03-12    0.023780
2016-03-13    0.008894
2016-03-14    0.012600
2016-03-15    0.015874
2016-03-16    0.016450
2016-03-17    0.028083
2016-03-18    0.007350
2016-03-19    0.015832
2016-03-20    0.020650
2016-03-21    0.020630
2016-03-22    0.021371
2016-03-23    0.018530
2016-03-24    0.019765
2016-03-25    0.019209
2016-03-26    0.016800
2016-03-27    0.015647
2016-03-28    0.020877
2016-03-29    0.022359
2016-03-30    0.024768
2016-03-31    0.023780
2016-04-01    0.022791
2016-04-02    0.024933
2016-04-03    0.025200
2016-04-04    0.024480
2016-04-05    0.124786
2016-04-06    0.221799
2016-04-07    0.131931
Name: last_seen, dtype: float64

We can see that most ads (47%) were seen last between 2016-04-05 and 2016-04-07, which are the most recent dates in the column.

Let's undestand now the distribution of **registration_year**.

In [371]:
autos['registration_year'].describe()

count    48571.000000
mean      2004.754257
std         88.638548
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [372]:
unique_years = autos['registration_year'].unique()
print(type(unique_years))
print(unique_years)

<class 'numpy.ndarray'>
[2004 1997 2009 2007 2003 2006 1995 1998 2000 2017 2010 1999 1982 1990
 2015 2014 1996 1992 2002 2012 2011 2005 2008 1985 2016 1994 1986 2001
 2018 2013 1972 1993 1988 1989 1973 1967 1976 4500 1987 1991 1983 1960
 1969 1950 1978 1980 1984 1963 1977 1961 1968 1934 1965 1971 1966 1979
 1981 1970 1974 1910 1975 5000 4100 2019 1956 9999 6200 1964 1959 1958
 1800 1948 1931 1943 1941 1962 1927 1937 1929 1000 1957 1952 1111 1955
 1939 8888 1954 1938 2800 5911 1953 1951 4800 1001 9000]


Let's sort the unique_years in order to identify the outliers.

In [373]:
sorted_unique_years = np.sort(unique_years)
print(sorted_unique_years)

[1000 1001 1111 1800 1910 1927 1929 1931 1934 1937 1938 1939 1941 1943
 1948 1950 1951 1952 1953 1954 1955 1956 1957 1958 1959 1960 1961 1962
 1963 1964 1965 1966 1967 1968 1969 1970 1971 1972 1973 1974 1975 1976
 1977 1978 1979 1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990
 1991 1992 1993 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004
 2005 2006 2007 2008 2009 2010 2011 2012 2013 2014 2015 2016 2017 2018
 2019 2800 4100 4500 4800 5000 5911 6200 8888 9000 9999]


We can see some odd values, like 1000 and 9999. Besides that, any vehicle with a registration year above 2016 is definitely inaccurate, because a car can't be registered after the listing was seen.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval.

In [374]:
year_interval = autos[(autos['registration_year'] < 1900) 
                       |
                       (autos['registration_year'] > 2016)
                     ]
#print(year_interval.head())
nr_years = year_interval.shape[0]
print('{} listings out of the interval 1900 - 2016'.format(nr_years))

1884 listings out of the interval 1900 - 2016


In [375]:
print(autos.shape)
autos = autos[autos['registration_year'].between(1900,2016,inclusive=True)]
print(autos.shape)

(48571, 20)
(46687, 20)


We have just removed 1884 lines of our dataset which contained non-sense years. By calling the `describe()` method again, we can see that now the values make sense. 

In [376]:
autos['registration_year'].describe()

count    46687.000000
mean      2002.909782
std          7.186105
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [377]:
autos['registration_year'].value_counts(normalize=True)

2000    0.067599
2005    0.062887
1999    0.062073
2004    0.057896
2003    0.057811
2006    0.057211
2001    0.056461
2002    0.053248
1998    0.050614
2007    0.048772
2008    0.047444
2009    0.044681
1997    0.041789
2011    0.034763
2010    0.034035
1996    0.029409
2012    0.028059
1995    0.026281
2016    0.026131
2013    0.017200
2014    0.014201
1994    0.013473
1993    0.009103
2015    0.008396
1992    0.007947
1990    0.007432
1991    0.007261
1989    0.003727
1988    0.002892
1985    0.002056
          ...   
1966    0.000471
1976    0.000450
1969    0.000407
1975    0.000386
1965    0.000364
1964    0.000257
1963    0.000171
1959    0.000129
1961    0.000129
1910    0.000107
1956    0.000086
1958    0.000086
1937    0.000086
1962    0.000086
1950    0.000064
1954    0.000043
1941    0.000043
1951    0.000043
1934    0.000043
1957    0.000043
1955    0.000043
1953    0.000021
1943    0.000021
1929    0.000021
1939    0.000021
1938    0.000021
1948    0.000021
1927    0.0000

Let's use aggregation to understand how the **price** and **brand** columns relate.

In [379]:
unique_brands = autos['brand'].unique()
print(unique_brands)

['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'sonstige_autos' 'opel' 'mazda' 'porsche' 'mini'
 'toyota' 'dacia' 'nissan' 'jeep' 'saab' 'volvo' 'mitsubishi' 'jaguar'
 'fiat' 'skoda' 'subaru' 'kia' 'citroen' 'chevrolet' 'hyundai' 'honda'
 'daewoo' 'suzuki' 'trabant' 'land_rover' 'alfa_romeo' 'lada' 'rover'
 'daihatsu' 'lancia']


In [390]:
normalized_brand = autos['brand'].value_counts(normalize=True, 
                                               ascending=False)
print(normalized_brand)

volkswagen        0.211279
bmw               0.110052
opel              0.107567
mercedes_benz     0.096451
audi              0.086555
ford              0.069912
renault           0.047144
peugeot           0.029837
fiat              0.025639
seat              0.018271
skoda             0.016407
nissan            0.015272
mazda             0.015186
smart             0.014158
citroen           0.014008
toyota            0.012702
hyundai           0.010024
sonstige_autos    0.009853
volvo             0.009146
mini              0.008760
mitsubishi        0.008225
honda             0.007839
kia               0.007068
alfa_romeo        0.006640
porsche           0.006126
suzuki            0.005933
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002270
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001499
trabant           0.001392
r

We'll aggregate only the brands that have over 5% of the total values.

In [393]:
top_normalized_brands = normalized_brand[normalized_brand > 0.05]
print(top_normalized_brands)

volkswagen       0.211279
bmw              0.110052
opel             0.107567
mercedes_benz    0.096451
audi             0.086555
ford             0.069912
Name: brand, dtype: float64


Let's do the aggregation with the following brands:

- Volkswagen
- BMW
- Opel
- Mercedes Benz
- Audi
- Ford

In [394]:
top_brands = top_normalized_brands.index
print(top_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [404]:
brand_mean_price = {}
brand_mean_mileage = {}

for brand in top_brands:
    
    selected_brand = autos[autos['brand'] ==  brand]
    
    brand_prices = selected_brand['price']
    mean_price = brand_prices.mean()
    brand_mean_price[brand] = mean_price
    
    brand_mileages = selected_brand['odometer_km']
    mean_mileage = brand_mileages.mean()
    brand_mean_mileage[brand] = mean_mileage

print(brand_mean_price)

{'volkswagen': 5604.071269261963, 'ford': 4054.6930147058824, 'opel': 2975.2419354838707, 'audi': 9336.687453600594, 'mercedes_benz': 8628.450366422385, 'bmw': 8571.480147917478}


As expected, the mean prices are higher for brands like Audi, BMW and Mercedes Benz and lower for Opel, Ford and Volkswagen.

Let's convert both dictionaries to series objects using the series constructor.

In [407]:
bmp_series = pd.Series(brand_mean_price)
print(bmp_series, end='\n\n')

bmm_series = pd.Series(brand_mean_mileage)
print(bmm_series)

audi             9336.687454
bmw              8571.480148
ford             4054.693015
mercedes_benz    8628.450366
opel             2975.241935
volkswagen       5604.071269
dtype: float64

audi             129157.386785
bmw              132575.905021
ford             124266.237745
mercedes_benz    130788.363313
opel             129310.035842
volkswagen       128711.476075
dtype: float64


Now we can create a single-column dataframe from the bmp series object.

In [410]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

mean_price
audi           9336.687454
bmw            8571.480148
ford           4054.693015
mercedes_benz  8628.450366
opel           2975.241935
volkswagen     5604.071269

In [411]:
df['mean_mileage'] = bmm_series
df

mean_price   mean_mileage
audi           9336.687454  129157.386785
bmw            8571.480148  132575.905021
ford           4054.693015  124266.237745
mercedes_benz  8628.450366  130788.363313
opel           2975.241935  129310.035842
volkswagen     5604.071269  128711.476075

From the table above, we can see that all mean mileages fall within a short interval (between 128,000 km and 132,000). 

Thus, we can say that the difference in the mean prices are caused by other factors and further investigation would be required in order to identify those factors.